runtime/runtime/src/pipelining.rs

```diff
@@ -0,0 +1,319 @@
+                Action::NonrefundableStorageTransfer(_) => {}
+            }
+        }
+        return any_function_calls;
+    }
+
+    /// Obtain the prepared contract for the provided receipt.
+    ///
+    /// If the contract is currently being prepared this function will block waiting for the
+    /// preparation to complete.
+    ///
+    /// If the preparation hasn't been started yet (either because it hasn't been scheduled for any
+    /// reason, or because the pipeline didn't make it in time), this function will prepare the
+    /// contract in the calling thread.
+    pub(crate) fn get_contract(
+        &self,
+        receipt: &Receipt,
+        code_hash: CryptoHash,
+        action_index: usize,
+        view_config: Option<ViewConfig>,
+    ) -> Box<dyn PreparedContract> {
+        let account_id = receipt.receiver_id();
+        let action = match receipt.receipt() {
+            ReceiptEnum::Action(r) | ReceiptEnum::PromiseYield(r) => r
+                .actions
+                .get(action_index)
+                .expect("indexing receipt actions by an action_index failed!"),
+            ReceiptEnum::Data(_) | ReceiptEnum::PromiseResume(_) => {
+                panic!("attempting to get_contract with a non-action receipt!?")
+            }
+        };
+        let Action::FunctionCall(function_call) = action else {
+            panic!("referenced receipt action is not a function call!");
+        };
+        let key = PrepareTaskKey { receipt_id: receipt.get_hash(), action_index };
+        let Some(task) = self.map.get(&key) else {
+            let gas_counter = self.gas_counter(view_config.as_ref(), function_call.gas);
+            return prepare_function_call(
+                &self.storage,
+                self.contract_cache.as_deref(),
+                &self.chain_id,
+                self.protocol_version,
+                Arc::clone(&self.config.wasm_config),
+                gas_counter,
+                code_hash,
+                &account_id,
+                &function_call.method_name,
+            );
+        };
+        let mut status_guard = task.status.lock().unwrap();
+        loop {
+            let current = std::mem::replace(&mut *status_guard, PrepareTaskStatus::Working);
+            match current {
+                PrepareTaskStatus::Pending => {
+                    let gas_counter = self.gas_counter(view_config.as_ref(), function_call.gas);
+                    let contract = prepare_function_call(
+                        &self.storage,
+                        self.contract_cache.as_deref(),
+                        &self.chain_id,
+                        self.protocol_version,
+                        Arc::clone(&self.config.wasm_config),
+                        gas_counter,
+                        code_hash,
+                        &account_id,
+                        &function_call.method_name,
+                    );
+                    *status_guard = PrepareTaskStatus::Finished;
```

Comment:
Nit: set to finished before actually calling `prepare_function_call`, and release the status guard, so that the mutex is held for less time? It might help if the prefetcher thread is coming to prefetch this contract just when the compute thread is handling it.